### ISSUE TITLE
SWIG 3.0.5 broken?

### ISSUE THREAD

**Issue by swails** [pandegroup/openmm#879]
I'm getting some strange behavior in SWIG 3.0.5 on my Mac (OS X 10.10.3).  In particular, some of the wrapped classes no longer accept ``*args`` as their argument list -- it is set to whatever is in the actual header file.  Some of the custom SWIG code depends on ``*args``, though, like `stripUnits`.  This causes all of the Python CHARMM tests to fail in `System.addParticle` and all Amber tests to fail in `System.addForce`.

In the SWIG-generated `openmm.py` file, this is where the Amber files are failing:

```Python
    def addForce(self, force):
        """
        addForce(System self, Force force) -> int

        Add a Force to the System. The Force should have been created on the heap with the "new" operator. The System takes over ownership of it, and deletes the Force when the System itself is deleted.
           Parameters:
            - force a pointer to the Force object to be added
        """
        if not args[0].thisown:
           s = "the %s object does not own its corresponding OpenMM object" \
              % args[0].__class__.__name__
           raise Exception(s)


        val = _openmm.System_addForce(self, force)
        args[0].thisown=0


        return val
```

This clearly throws a `NameError` when trying to access `thisown`.  Likewise, for the CHARMM files:

```Python
    def addParticle(self, mass):
        """
        addParticle(System self, double mass) -> int

        Add a particle to the System. If the mass is 0, Integrators will ignore the particle and not modify its position or velocity. This is most often used for virtual sites, but can also be used as a   way to prevent a particle from moving.
           Parameters:
            - mass the mass of the particle (in atomic mass units)
        """
        try: args=stripUnits(args)
        except UnboundLocalError: pass


        return _openmm.System_addParticle(self, mass)
```

Since `addParticle` is called with mass being a `unit.Quantity` instance, the `return` line fails with a `TypeError` (not a `double`).

Is anyone else seeing this behavior?  Is it a bad CMake setting?

**Comment by swails** [comment]
So I checked out version 6.1 and compiled that.  I got the same issue, so it's definitely a SWIG 3.0.5 thing.  I don't know how to tell SWIG to go back to using `*args` for all of the wrapped call signatures.

Am I the only one seeing this?

**Comment by peastman** [comment]
I've been going through the SWIG 3 documentation trying to find anything related to this.  At http://www.swig.org/Doc3.0/Python.html#Python_nn67 it says:

"As alluded to above SWIG will generate all the function and method proxy wrappers with just "\*args" (or "\*args, \*\*kwargs" if the -keyword option is used) for a parameter list and will then sort out the individual parameters in the C wrapper code."

That's certainly what it _used_ to do.  Maybe they've changed it and not updated the documentation?  On the other hand, at http://www.swig.org/Doc3.0/Python.html#Python_nn74 it talks about using a different method to pass arguments.  It's discussed in the context of Python 3, so perhaps it only happens then?  And it doesn't mention anything about a way of turning it off.

**Comment by swails** [comment]
I filed an issue with swig's github repo and am investigating the code side of swigs changes.

**Comment by swails** [comment]
So I found the offending commit in swig: swig/swig@fdc6bbeda387fb49ab7dcdbb0f17e4645db3115c.  Let's see what they say in swig/swig#387 and go from there.  In the meantime I can work around via an older SWIG.

**Comment by swails** [comment]
Ugh.  Reading that commit log in the swig git log suggests that they consider always using *args to be a bug, not a feature.  It looks like we might have to beef up the Python script that creates the SWIG input file a lot (i.e., we can't simply use
```Python
    def whatever(self, *args):
        try: stripUnits(args)
        except UnboundLocalError: pass
```
in the beginning of every function).

**Comment by swails** [comment]
OK -- looks like we need to use the `"compactdefaultargs"` feature for every method.  I'll do this in a PR.

**Comment by swails** [comment]
Should be fixed in #882.  I've verified that this fixes SWIG 3 on my Mac.
